ДЗ5 готово

EvgeniiKovalev opened the pull request:
Есть метод только отладки - делает печать массива, для удобства. 
Дополнительно к осн. заданию написал вариант без цикла в setOfArray
Проверил аналог foreach в increaseOfElementArray - не заработал, предположил что 
цикл foreach, элементы перебирает, но изменить в таком цикле их не получается, изменения не сохраняются в массиве, скорее всего потому что item лок. переменная

EvgeniiKovalev committed:
ДЗ5 готово

FlameXander commented on src/main/java/ru/otus/java/basic/homeworks/homework5/AppHw5.java:
```diff
@@ -0,0 +1,91 @@
+            array[i] += decimalNumber;
+        }
+
+        //аналог цикла foreach, элементы перебирает, но изменить в таком цикле их не получается, изменения не сохраняются в массиве, скорее всего потому что item лок. переменная
```
Все верно, меняя эту переменную никак нельзя повлиять на массив, так как они не связаны. Переменная просто последовательно инициализируется значением каждого элемента массива

FlameXander commented on src/main/java/ru/otus/java/basic/homeworks/homework5/AppHw5.java:
```diff
@@ -0,0 +1,91 @@
+        //System.out.println("печать массива: " + Arrays.toString(array));
+
+        //вариант с циклом
+        for (int i = 0; i < array.length; i++) {
```
Чтобы не "напрягать" цикл лишними проверками лучше так сделать (+уйдет необходимость вчитываться в тернарный оператор):
```
System.out.print("Печать массива: [");
for (int i = 0; i < array.length; i++) {
    System.out.print(array[i] + ", ");
}
System.out.println("]");
```

EvgeniiKovalev committed:
решены задачи со *
упорядочил main

EvgeniiKovalev committed:
исправил snake_case на camelCase

FlameXander commented on src/main/java/ru/otus/java/basic/homeworks/homework5/AppHw5.java:
```diff
@@ -4,16 +4,177 @@
+            maxLengthArrays = Math.max(maxLengthArrays, i.length);
+        }
+        int[] resArr = new int[maxLengthArrays]; //инициализирован нулями
+        for (int i = 0; i < maxLengthArrays; i++) {
```
Нет смысла заставлять джаву генерировать лишнюю пачку исключений, упрощаем:
```
for (int[] ints : arr) {
    for (int i = 0; i < ints.length; i++) {
        resArr[i] += ints[i];
    }
}
```

FlameXander commented on src/main/java/ru/otus/java/basic/homeworks/homework5/AppHw5.java:
```diff
@@ -0,0 +1,255 @@
+    //    Реализуйте метод, проверяющий что есть "точка" в массиве, в которой сумма левой и правой части равны.
+//    Точка находится между элементами.
+//    Пример: { 1, 1, 1, 1, 1, | 5 }, { 5, | 3, 4, -2 }, { 7, 2, 2, 2 }, { 9, 4 }
+    public static void FindDotArray(int[] arr) {
```
camelCase

FlameXander replied:
Решение правильное, но можно было бы оптимизировать и не делать на каждом шаге пересчет каждой из сторон

EvgeniiKovalev replied:
как написать кратко опишу  -  на второй итерации уже можно перейти к оптимальномук вычислению сумм (левой и правой частей массива) от левой суммы вседа отнимаем текущее число, к правой прибавляем. 
Если можно без оптимизации сдать то сдал бы без оптимизации, т.к. понятно как это делать.

FlameXander commented on src/main/java/ru/otus/java/basic/homeworks/homework5/AppHw5.java:
```diff
@@ -0,0 +1,255 @@
+
+    //    Реализуйте метод, проверяющий что все элементы массива идут в порядке убывания или
+//    возрастания (по выбору пользователя)
+    public static String arrayIsOrdered(int[] arr, boolean direction) {
```
Названия и тип аргументов должны быть такими чтобы их не приходилось объяснять в комментариях. Тут конечно лучше енум использовать (но это будущая тема). boolean checkIsAsc - проверяем что возрастающий порядок

FlameXander commented on src/main/java/ru/otus/java/basic/homeworks/homework5/AppHw5.java:
```diff
@@ -0,0 +1,255 @@
+
+    //    Реализуйте метод, проверяющий что все элементы массива идут в порядке убывания или
+//    возрастания (по выбору пользователя)
+    public static String arrayIsOrdered(int[] arr, boolean direction) {
```
Тип метода тоже должен быть булеан, так как от возвращаемого текста мало толку

FlameXander replied:
Цикл фор сильно бы упростил код

EvgeniiKovalev replied:
на 4 строки сократился код метода, сложность таже, как цикломатическая, так и все прочие.
Видимо дело не в цикле, а алгоритм менять нужно кардинально, пока не придумал

FlameXander replied:
Давайте просто приведу пример упрощения через фор:
```
    public static boolean isOrdered(int[] array, boolean checkIsAsc) {
        if (array.length < 2) {
            return true;
        }
        for (int i = 0; i < array.length - 1; i++) {
            if (array[i + 1] != array[i] && array[i + 1] > array[i] != checkIsAsc) {
                return false;
            }
        }
        return true;
    }
```
Проверка на то что массив или не убывающий, или не возрастающий (т.е. соседние элементы могут быть равны и это не помешает проверке)

FlameXander commented on src/main/java/ru/otus/java/basic/homeworks/homework5/AppHw5.java:
```diff
@@ -0,0 +1,255 @@
+
+    //    Реализуйте метод, “переворачивающий" входящий массив
+//    Пример: { 1 2 3 4 } => { 4 3 2 1 }
+    public static void arrayReverse(int[] arr) {
```
Тут точно массив перевернется? Или перевернется дважды и вернется почти в исходное состояние?

EvgeniiKovalev replied:
Конечно нет, поправил, но мыслил все равно в верном направлении

EvgeniiKovalev committed:
пофиксил кучу замечаний, какие именно - написал в обсуждении ПР на гитхабе

EvgeniiKovalev committed:
пофиксил arrayReverse
не проходил тест на массиве с нечетным количеством элементов
и проходил с четным
